### ISSUE TITLE
Can't login. Line 151 in Client.php to blame.

### ISSUE THREAD

**Issue by ericfri** [Art-of-WiFi/UniFi-API-client#31]
On my current setup, I was having problems logging in and I've traced the issue to line 151 in Client.php. The line is as follows:

`preg_match_all('|Set-Cookie: (.*);|U', substr($content, 0, $header_size), $results);`

In my instance, the reponse is all lowecase, so 'Set-Cookie' doesn't match 'set-cookie'

Changing the line to 
`preg_match_all('|set-cookie: (.*);|u', substr($content, 0, $header_size), $results);`
fixed the issue in my case. But can this line to changed to accept any case?

**Comment by malle-pietje** [comment]
I have never seen or heard of this happening before (and the API client has been downloaded more than 7000 times...). Until I see actual proof this is needed, I don't plan on changing this piece of code.

One more thing to look into is your exact PHP and cURL versions.

If you want, you can share the controller details with credentials for an account limited to an empty test site offline. Let me know if you want me to try this out.

Alternatively, you can simply fork the repo and adapt the code to your use case.

**Comment by ericfri** [comment]
PHP is version 7.2.13.
cURL is version 7.62.0

I'll create an account and send you the details to look at.

**Comment by malle-pietje** [comment]
Hmm, that's not too different from what our Search tool demo runs off:
![searchtooldetails](https://user-images.githubusercontent.com/12016131/49960201-f1231280-ff0f-11e8-842b-23f229e65561.PNG)

https://search-tool-demo.artofwifi.net/

Yes please, please send that to info AT artofwifi DOT net (will remove this email address as soon as I receive your email)

**Comment by malle-pietje** [comment]
I've connected to your controller with the latest version of the API browser (1.0.34) and API client class (1.1.36) and all looks fine here.

This is the debug output from the login response:
```
HTTP/1.1 200 OK
Server: nginx
Date: Thu, 13 Dec 2018 19:02:24 GMT
Content-Type: application/json;charset=UTF-8
Content-Length: 41
Connection: keep-alive
Set-Cookie: unifises=xxxxxxxxxxxxxxxxxxxxxxxxxx; Path=/; Secure; HttpOnly
Set-Cookie: csrf_token=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx; Path=/; Secure
X-Frame-Options: DENY
Strict-Transport-Security: max-age=31536000
X-Frame-Options: DENY

{ "data" : [ ] , "meta" : { "rc" : "ok"}}
```
Although my suspicions something else _could_ be mangling the headers are confirmed (an nginx reverse proxy). Not sure though why I don't see the same symptoms at my end.

Maybe its related to your cURL version?

**Comment by ericfri** [comment]
Weird. The output I get is this:
`
----------RESPONSE-----------
HTTP/2 200 
server: nginx
date: Thu, 13 Dec 2018 19:11:09 GMT
content-type: application/json;charset=UTF-8
content-length: 41
set-cookie: unifises=w704lUYH4BBBsnyavNJIGnJ57zmhnFlY; Path=/; Secure; HttpOnly
set-cookie: csrf_token=GcL6nMZQ0gbDbGfvzFXi3YzS806Fpavg; Path=/; Secure
x-frame-options: DENY
strict-transport-security: max-age=31536000
x-frame-options: DENY

{ "data" : [ ] , "meta" : { "rc" : "ok"}}
`

**Comment by malle-pietje** [comment]
Maybe nginx is returning different headers based on the destination? I have never used nginx as a reverse proxy so can't talk from experience...

**Comment by malle-pietje** [comment]
Quick check could be to use a different server with an older PHP and cURL version to execute the requests from? An RPi will do:wink:

**Comment by ericfri** [comment]
I can't change the server. It's a production server we use for our web hosting clients. The client is all cPanel and the UniFi server is a stand alone VM.

**Comment by malle-pietje** [comment]
Just as a test to isolate the root-cause. I believe BTW with cPanel you should be able to switch PHP versions, but not sure. It was a while ago I used cPanel...

**Comment by malle-pietje** [comment]
OK, in that case, I'd look into the nginx configuration.

**Comment by malle-pietje** [comment]
Don't know about PHP FPM though.

**Comment by ericfri** [comment]
Well using CGI vs FPM and PHP versions 7.0 and 7.2 didn't change anything. cURL is the same across all the tests, so maybe that's the issue.
Either way, wouldn't be pretty easy to make the preg_match_all case insensitive? Then whatever is causing the issue wouldn't even be an issue any more.

**Comment by malle-pietje** [comment]
If you have a suggestion for the code change, please share and I will have a look into it. I might also look at using cURL to extract the cookies and other headers, not sure how version dependent that is though.

**Comment by ericfri** [comment]
Changing the line from
`preg_match_all('|Set-Cookie: (.*);|U', substr($content, 0, $header_size), $results);`
to
`preg_match_all('|Set-Cookie: (.*);|Ui', substr($content, 0, $header_size), $results);`

seems to work for me. I just added the 'i' after the 'U'

**Comment by malle-pietje** [comment]
Thanks, I’ll look this.

**Comment by malle-pietje** [comment]
Looks like lowercase header field names are allowed per the RFC:
https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2

**Comment by malle-pietje** [comment]
This is now fixed with Release v1.1.38: https://github.com/Art-of-WiFi/UniFi-API-client/releases/tag/v1.1.38
